Support injection of AgroalPoolInterceptor

barreiro opened the pull request:
This is fix for #5381 alternative to #5408 

Agroal 1.8 introduces a new interceptor mechanism suited to be used by applications. This PR allows CDI to detect and inject these interceptors into the right datasources.

mkouba reviewed (commented):
LGTM.

The only problem I can see is that if a `AgroalPoolInterceptor` bean with scope `@Dependent` and a lifecycle callback (e.g. `@PreDestroy`) is used the callback will never be called.  

A solution could be to (1) destroy the bean via `Instance#destroy()` or (2) use `io.quarkus.arc.InjectableInstance` instead, store the ``InstanceHandle``s  and call `InstanceHandle#close()` when needed (i.e. when the corresponding DS is destroyed).

gsmet commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/InstanceImpl.java:
```diff
@@ -138,7 +138,7 @@ public void destroy(Object instance) {
     @Override
     public Iterable<InstanceHandle<T>> handles() {
-        return (Iterable<InstanceHandle<T>>) beans().stream().map(this::getHandle).iterator();
+        return beans().stream().map(bean -> (InstanceHandle<T>) getHandle(bean))::iterator;
```
Please no lambdas in runtime code.

barreiro commented:
@mkouba I addressed your comment on the life-cycle of the interceptors using solution 2). There was a `ClassCastException` on `InstanceImpl#handles()` that I had to fix, so please review that. Thanks!

mkouba commented:
@barreiro I don't know much about the agroal extension and the `AbstractDataSourceProducer` class but it seems that the class itself is used as a base of a regular class bean and thus supports injection. Therefore, you could just do `@Inject Instance<AgroalPoolInterceptor>` and in this case, the destruction of `@Dependent` instances created through this construct is performed automatically when the bean is destroyed. I'm sorry that I didn't spot this before :-(.

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/InstanceImpl.java:
```diff
@@ -138,7 +138,7 @@ public void destroy(Object instance) {
     @Override
     public Iterable<InstanceHandle<T>> handles() {
-        return (Iterable<InstanceHandle<T>>) beans().stream().map(this::getHandle).iterator();
+        return (Iterable<InstanceHandle<T>>) (Iterable) beans().stream().map(this::getHandle)::iterator;
```
And pls remove this cast.

barreiro committed:
Support injection of AgroalPoolInterceptor

barreiro commented:
Updated to `@Inject @All Instance<AgroalPoolInterceptor>`. Thanks for the suggestion.

mkouba commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java:
```diff
@@ -65,6 +71,10 @@
 
+    @Inject
+    @Any
+    public Instance<AgroalPoolInterceptor> agroalPoolInterceptors;
```
Is there a good reason to have this field `public` instead of `protected` or `package-private`? And the same applies to other injection points in this class.

barreiro replied:
No idea why. Changed all the injection points to `package-private`

mkouba commented:
@barreiro Looks good now but there are some conflicts...

barreiro commented:
rebased and resoved conflicts

barreiro commented:
rebased once again

gsmet commented:
@Sanne I suppose you want to review that one as you were deep in the discussions around that?

Sanne commented:
thanks @gsmet - I'll have a look now

Sanne reviewed (approved):
Great, thanks all!
